public_html/wp-content/themes/pattern-directory/src/components/category-context-bar/index.js

```diff
@@ -1,30 +1,82 @@
 /**
- * External dependencies
+ * WordPress dependencies
  */
 import { useEffect, useRef, useState } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+import { getQueryArg } from '@wordpress/url';
 
-function CategoryContextBar( { message, title, links } ) {
+/**
+ * Internal dependencies
+ */
+import { useRoute } from '../../hooks';
+import { getCategoryFromPath } from '../../utils';
+import { getContextMessage } from './messaging';
+import { store as patternStore } from '../../store';
+
+function CategoryContextBar() {
+	const { path } = useRoute();
 	const [ height, setHeight ] = useState( 0 );
+	const [ context, setContext ] = useState( {} );
 	const innerRef = useRef( null );
 
+	const { isAllCategory, category, isLoadingPatterns, patterns } = useSelect( ( select ) => {
+		const {
+			getCategoryBySlug,
+			getPatternsByQuery,
+			isLoadingPatternsByQuery,
+			getCurrentQuery } = select( patternStore );
+		const categorySlug = getCategoryFromPath( path );
+		const _category = getCategoryBySlug( categorySlug );
+		const query = getCurrentQuery();
+
+		return {
+			isAllCategory: _category && _category.id === -1,
+			isLoadingPatterns: isLoadingPatternsByQuery( query ),
+			patterns: query ? getPatternsByQuery( query ) : [],
+			category: _category,
+		};
+	}, [ path ] );
+
 	useEffect( () => {
-		if ( message ) {
-			setHeight( innerRef.current.offsetHeight );
-		} else {
-			setHeight( 0 );
+		if ( ! category || isLoadingPatterns ) {
+			return;
 		}
-	}, [ message ] );
+
+		const searchTerm = getQueryArg( path, 'search' );
+
+		let _context = {};
+
+		// Use the category count as default since it has the count of all the associated patterns
+		let count = category.count;
+
+		// If we have a search term use the pattern results length.
+		// Not: This is okay until we start using paging.
+		if ( searchTerm ) {
+			count = patterns.length;
+		}
+
+		if ( ! isAllCategory || searchTerm ) {
+			_context = getContextMessage( count, category.name, searchTerm );
+		}
+
+		setContext( _context );
+	}, [ isLoadingPatterns, patterns ] );
```

Comment:
Good observation.

Technically, since patterns are not queried until we have built the query and saved it to the state, and building the query requires that categories have already loaded... `category` _should_ always be defined before patterns, at least based on the fact that we load this component immediately. With those assumptions in place, we can save a few renders by not watching the `category` updates. 

Should the component be conditionally rendered (or its parent) it would most likely need to observe `category` changes. Now, the likelihood that this component will ever contain children that could have expensive renders is very low, and thinking in terms of performance versus portability, it may make more sense for us to additionally track `category` to protect against upstream re-structuring. 

Lastly, and most importantly, I've totally overthought this 🤪.